libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/NavigationRouteProcessor.kt

```diff
@@ -0,0 +1,141 @@
+package com.mapbox.services.android.navigation.v5.internal.navigation
+
+import com.mapbox.api.directions.v5.models.DirectionsRoute
+import com.mapbox.api.directions.v5.models.LegStep
+import com.mapbox.api.directions.v5.models.RouteLeg
+import com.mapbox.geojson.Geometry
+import com.mapbox.geojson.Point
+import com.mapbox.navigator.NavigationStatus
+import com.mapbox.navigator.RouteState
+import com.mapbox.services.android.navigation.v5.routeprogress.CurrentLegAnnotation
+import com.mapbox.services.android.navigation.v5.routeprogress.RouteProgress
+import com.mapbox.services.android.navigation.v5.routeprogress.RouteProgressStateMap
+
+internal class NavigationRouteProcessor {
+    private val ONE_INDEX = 1
+    private val ONE_SECOND_IN_MILLISECONDS = 1000.0
+    private val FIRST_BANNER_INSTRUCTION = 0
+
+    private val progressStateMap = RouteProgressStateMap()
+    private var previousRouteProgress: RouteProgress? = null
+    private var previousStatus: NavigationStatus? = null
+    private var route: DirectionsRoute? = null
+    private var currentLeg: RouteLeg? = null
+    private var currentStep: LegStep? = null
+    private var currentStepPoints: List<Point>? = null
+    private var upcomingStepPoints: List<Point>? = null
+    private var currentLegAnnotation: CurrentLegAnnotation? = null
+    private var routeGeometryWithBuffer: Geometry? = null
+
+    fun buildNewRouteProgress(navigator: MapboxNavigator, status: NavigationStatus, route: DirectionsRoute): RouteProgress? {
+        previousStatus = status
+        updateRoute(route, navigator)
+        return buildRouteProgressFrom(status, navigator)
+    }
+
+    fun updatePreviousRouteProgress(routeProgress: RouteProgress?) {
+        previousRouteProgress = routeProgress
+    }
+
+    fun retrievePreviousRouteProgress() =
+            previousRouteProgress
+
+    fun retrievePreviousStatus() =
+            previousStatus
+
+    private fun updateRoute(route: DirectionsRoute, navigator: MapboxNavigator) {
+        if (this.route == null || this.route != route) {
+            this.route = route
+            routeGeometryWithBuffer = navigator.retrieveRouteGeometryWithBuffer()
+        }
+    }
+
+    private fun buildRouteProgressFrom(status: NavigationStatus, navigator: MapboxNavigator): RouteProgress? {
```

Comment:
See comment above re: `RouteProgress?` should be non-nullable

Reply:
same as above